.changeset/tall-ties-jog.md

```diff
@@ -0,0 +1,54 @@
+---
+'@commercetools-frontend/application-shell': minor
+---
+
+New utility functions to configure HTTP requests when using custom HTTP clients.
+
+By default Custom Applications provide [preconfigured HTTP clients](https://docs.commercetools.com/custom-applications/development/data-fetching) for GraphQL and REST API requests.
+
+However, you could use any other HTTP client of your choice, for example `fetch`, `axios`, `swr`, etc.
+
+The main problem with that is the fact that you need to configure these clients on your own, in particular regarding the HTTP headers that should be sent with every request.
+
+This is now possible using some dedicated utility functions. See example usage below:
+
+```js
+import createHttpUserAgent from '@commercetools/http-user-agent';
+import {
+  buildApiUrl,
+  createHttpClientOptions,
+  executeHttpClientRequest,
+} from '@commercetools-frontend/application-shell';
+
+const userAgent = createHttpUserAgent({
+  name: 'fetch-client',
+  version: '2.6.0',
+  libraryName: window.app.applicationName,
+  contactEmail: 'support@my-company.com',
+});
+
+// Simple example using `fetch`.
+const data = await executeHttpClientRequest(async (headers) => {
```

Comment:
Personally, I found a little difficult to understand this `headers` object we receive here and it's needed to be passed when creating the http client options we provide to `fetch` function.

After reading the code, I understand is needed for the token renewal to work properly.
However, if we are going to provide a function to manage the whole renewal process, I think we could maybe hide that away from the consumers.
I'm thinking is something like this:
```
const data = await executeHttpClientRequest(userAgent, async (httpClientOptions) => {
  const res = await fetch(buildApiUrl('/proxy/ctp/channels'), {
    method: 'GET',
    ...httpClientOptions,
  });
  const data = await res.json();

  return {
    data,
    statusCode: res.status,
    getHeader: (key) => res.headers.get(key),
  };
});
```

This way, as a consumer I know the `executeHttpClientRequest` is providing me the options I should pass to `fetch` (or others) when sending the actual request.
If I need to add custom headers, I would expect something like this:
```
const data = await executeHttpClientRequest(userAgent, async (httpClientOptions) => {
  httpClientOptions.headers['X-Custom-Header'] = 'Custom-Value';

  const res = await fetch(buildApiUrl('/proxy/ctp/channels'), {
    method: 'GET',
    ...httpClientOptions,
  });
  const data = await res.json();

  return {
    data,
    statusCode: res.status,
    getHeader: (key) => res.headers.get(key),
  };
});
```

Perhaps we want to keep exporting the `createHttpClientOptions` for users if they want to use them in any other way but, if we do that, maybe we also would need to provide helpers to handle the session token renewal by exporting not only `oidc-storage.getSessionToken` but also a constant with the response header name `x-refreshed-session-token`.

The current proposal is also valid for me, but I wanted to share these thoughts.

Reply:
Yes the API of this function is not super simple. I'll take another look following your suggestions, thanks!

I think the important part is to "hide" the token renewal logic, as it would be weird to leave this up to the user. Keeping it an implementation details would allow us to manage changes to it internally without relying on users to migrate.

Reply:
@CarlosCortizasCT How about this? https://github.com/commercetools/merchant-center-application-kit/pull/2734/commits/daacd10f7de4bcd973d1ddc8b45fb0b23509802c

Reply:
I find it easier to understand.
The only thing I think we perhaps could update is the order of `executeHttpClientRequest`: since the first parameter is now optional and the second is required, I would switch them:
```
// BEFORE
function executeHttpClientRequest<Data>(
  config: TConfig = {},
  fetcher: TFetcher<Data>
): Promise<Data>

// AFTER
function executeHttpClientRequest<Data>(
  fetcher: TFetcher<Data>,
  config: TConfig = {}
): Promise<Data> 
```

Reply:
Function overloading didn't help much. I changed now the order of the arguments as you suggested: https://github.com/commercetools/merchant-center-application-kit/pull/2734/commits/a8872e47e1819a489099c76f25d145e55448a491

Thanks!